File: DataStructureTypeScript/DoubleLinkedList.ts
import DNode from './common/DoubleLinkedListNode'

export default class DoubleLinkedList<E> {
  private dummyHead: DNode<E> = new DNode(null, null);
  private dummyTail: DNode<E> = new DNode(null, null);
  private size: number = 0;

  constructor() {
    this.dummyHead.next = this.dummyTail;
    this.dummyTail.prev = this.dummyHead;
  }

  public getSize(): number {
    return this.size;
  }

  // 向链表头添加一个元素
  public addFirst(node: DNode<E>) {
    node.next = this.dummyHead.next;
    node.prev = this.dummyHead;
    this.dummyHead.next.prev = node;
    this.dummyHead.next = node;
    this.size++;
  }

  // 删除链表的一个元素
  public remove(node: DNode<E>): void {
    node.prev.next = node.next;
    node.next.prev = node.prev;
    this.size--;
  }

  // 删除链表的最后一个元素
  public removeLast(node: DNode<E>): DNode<E> {
    if (this.dummyTail.prev == this.dummyHead)
      return null;
    const last = this.dummyTail.prev;
    this.remove(last);
    return last;
  }
}

File: DataStructureJava/DoubleLinkedList.java
public class DoubleLinkedList<E> {

    /**
     * 双向链表的节点信息
     * 每个节点都包含指向前一个和后一个节点的指针
     * 存储于节点上的key->value键值对
     */
    private class Node {
        public Node prev, next;
        public E key, value;
        public Node(E key, E value) {
            this.key = key;
            this.value = value;
        }
    }

    // 定义一个虚拟的头结点和虚拟尾结点
    private Node dummyHead, dummyTail;
    private int size;

    public DoubleLinkedList() {
        dummyHead = new Node(null, null);
        dummyTail = new Node(null, null);
        dummyHead.next = dummyTail;
        dummyTail.prev = dummyHead;
    }

    // 在链表头部添加元素
    public void addFirst(Node node) {
        // 先处理新插入的节点的prev和next
        node.next = dummyHead.next;
        node.prev = dummyHead;
        // 在处理之前dummyHead的next和dummyHead.next.prev
        dummyHead.next.prev = node;
        dummyHead.next = node;
        size++;
    }

    // 删除双链表中的node节点
    public void remove(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
        size--;
    }

    // 移除双链表中最后一个节点，并返回删除的元素
    public Node remove() {
        if (dummyTail.prev == dummyHead)
            return null;
        Node last = dummyTail.prev;
        remove(last);
        return last;
    }
    
    // 获取链表中的节点个数
    public int getSize() {
        return size;
    }
}
